File: sdks/aperture-js/example/concurrency_limit.ts
import { ApertureClient } from "@fluxninja/aperture-js";
import inquirer from 'inquirer';

async function initializeApertureClient() {
    const answers = await inquirer.prompt([
        {
            type: 'input',
            name: 'address',
            message: 'Enter your organization\'s address:',
        },
        {
            type: 'input',
            name: 'apiKey',
            message: 'Enter the API key:',
        },
    ]);

    const apertureClient = new ApertureClient({
        address: answers.address,
        apiKey: answers.apiKey,
    });

    return apertureClient;
}

async function sendRequest(apertureClient: ApertureClient) {
  // START: CLStartFlow
  const flow = await apertureClient.startFlow("concurrency-limiting-feature", {
    labels: {
      user_id: "some_user_id",
    },
    grpcCallOptions: {
      deadline: Date.now() + 1000,
    },
  });
  // END: CLStartFlow

  // START: CLFlowShouldRun
  if (flow.shouldRun()) {
    console.log("Request accepted. Processing..." + flow.checkResponse());
  } else {
    console.log("Request rejected due to concurrency limit. Try again later.");
  }

  await flow.end();
  // END: CLFlowShouldRun
}

async function handleConcurrencyLimit(apertureClient: ApertureClient) {
  const requestsPerSecond = 5;
  const durationInSeconds = 1000;
  for (let i = 0; i < durationInSeconds; i++) {
    const requests = Array.from({ length: requestsPerSecond }, () =>
      sendRequest(apertureClient),
    );
    // sending requests in parallel to simulate concurrency
    await Promise.all(requests);

    // Wait 1 second before sending the next batch of requests
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }
}

async function main() {
  const apertureClient = await initializeApertureClient();
  await handleConcurrencyLimit(apertureClient);
}

main().catch((e) => console.error(e));

File: sdks/aperture-py/example/concurrency_limit.py
import asyncio

from aperture_sdk.client import ApertureClient, FlowParams

requests_per_second = 5
duration_in_seconds = 10

agent_address = ""
api_key = ""


async def initialize_aperture_client():
    aperture_client = ApertureClient.new_client(address=agent_address, api_key=api_key)
    return aperture_client


async def send_request(aperture_client):
    print("Sending request...")
    flow_params = FlowParams(
        explicit_labels={"user_id": "some_user_id"},
    )
    flow = aperture_client.start_flow(
        control_point="concurrency-limiting-feature",
        params=flow_params,
    )

    if flow.should_run():
        print("Request accepted. Processing...")
        # Simulate processing
        await asyncio.sleep(5)
    else:
        print("Request rejected due to concurrency limit. Try again later.")

    await flow.end()


async def handle_concurrency_limit(aperture_client):
    for _ in range(duration_in_seconds):
        tasks = [
            asyncio.create_task(send_request(aperture_client))
            for _ in range(requests_per_second)
        ]
        await asyncio.gather(*tasks)
        await asyncio.sleep(1)


async def main():
    aperture_client = await initialize_aperture_client()
    await handle_concurrency_limit(aperture_client)


if __name__ == "__main__":
    asyncio.run(main())
